modules/table/src/main/java/org/apache/ignite/internal/table/AbstractTableView.java

```diff
@@ -95,17 +101,40 @@ protected final <T> T sync(CompletableFuture<T> fut) {
      * @return Whatever the action returns.
      */
     protected final <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, KvAction<T> action) {
-        // TODO: IGNITE-20106 - retry if our request is rejected by the server due to a changed schema version.
+        return withSchemaSync(tx, null, action);
+    }
 
+    private <T> CompletableFuture<T> withSchemaSync(@Nullable Transaction tx, @Nullable Integer previousSchemaVersion, KvAction<T> action) {
         CompletableFuture<Integer> schemaVersionFuture = tx == null
                 ? schemaVersions.schemaVersionAtNow(tbl.tableId())
                 : schemaVersions.schemaVersionAt(((InternalTransaction) tx).startTimestamp(), tbl.tableId());
 
-        CompletableFuture<T> future = schemaVersionFuture.thenCompose(action::act);
+        CompletableFuture<T> future = schemaVersionFuture
+                .thenCompose(schemaVersion -> {
+                    return action.act(schemaVersion)
+                            .handle((BiFunction<T, Throwable, CompletableFuture<T>>) (res, ex) -> {
```

Comment:
In my proposal I also provided a way to get rid of the cast to `BiFunction`, shall we use it?